File: src/main/java/org/hanoi/HanoiGame.java
package org.hanoi;

import java.util.*;
import java.util.stream.Collectors;

public class HanoiGame
{
    private List<Stack<Integer>> towers = new ArrayList<>(3);

    public HanoiGame(Stack<Integer> initialTower, Integer towerCount)
    {
        towers.add(initialTower);
        for (int i = 0; i < towerCount - 1; i++)
            towers.add(new Stack<>());
    }

    public static HanoiGame DefaultGame()
    {
        var disks = new Integer[] {5, 4, 3, 2, 1};
        var initialTower = new Stack<Integer>();

        for (Integer disk : disks)
            initialTower.push(disk);

        return new HanoiGame(initialTower, 3);
    }

    private void pushDisk(Integer towerIndex, Integer disk)
    {
        Stack<Integer> tower = getTower(towerIndex);
        if (tower.empty() || tower.peek() > disk)
            tower.push(disk);
    }

    private Integer popDisk(Integer towerIndex)
    {
        Stack<Integer> tower = getTower(towerIndex);

        try {
            return tower.pop();
        } catch (Exception e) {
            return null;
        }
    }

    public void applyMove(HanoiMove move)
    {
        Integer disk = popDisk(move.from);
        pushDisk(move.to, disk);
    }

    public List<HanoiMove> getPossibleMoves()
    {
        List<HanoiMove> moves = new ArrayList<>();
        List<Integer> topDisks = getTopDisks();

        for (int i = 0; i < topDisks.size(); i++)
            for (int j = 0; j < topDisks.size(); j++)
                if (topDisks.get(i) != null && (topDisks.get(j) == null || topDisks.get(i) < topDisks.get(j)))
                    moves.add(new HanoiMove(i, j));

        return moves;
    }

    public Integer getTopDisk(Integer towerIndex)
    {
        Stack<Integer> tower = getTower(towerIndex);

        try {
            return tower.peek();
        } catch (Exception e) {
            return null;
        }
    }

    public List<Integer> getTopDisks()
    {
        List<Integer> tops = new ArrayList<>(towers.size());
        for (int i = 0; i < towers.size(); i++)
            tops.add(getTopDisk(i));

        return tops;
    }

    public Stack<Integer> getTower(Integer index)
    {
        return towers.get(index);
    }

    public boolean isComplete()
    {
        return towers.stream()
                .dropWhile(Stack::empty)
                .count() <= 1;
    }

    @Override
    public String toString() {
        return towers.stream().map(tower -> tower.toString().concat("\n")).collect(Collectors.joining());
    }
}

File: hanoi_game.py
from copy import deepcopy

from search import Problem

class HanoiMove:
    def __init__(self, from_tower, to_tower):
        self.from_tower = from_tower
        self.to_tower = to_tower

    def __str__(self):
        return "{} -> {}".format(self.from_tower, self.to_tower)


def default():
    return HanoiGame([[5, 4, 3, 2, 1], [], []])


class HanoiGame:
    def __init__(self, towers):
        self.towers = towers

    def push_disk(self, tower_index, disk):
        tower = self.get_tower(tower_index)
        if len(tower) == 0 or tower[-1] > disk:
            tower.append(disk)

    def pop_disk(self, tower_index):
        tower = self.get_tower(tower_index)
        if len(tower) > 0:
            return tower.pop()
        return None

    def apply_move(self, move):
        disk = self.pop_disk(move.from_tower)
        self.push_disk(move.to_tower, disk)

    def get_possible_moves(self):
        moves = []
        top_disks = self.get_top_disks()

        for i in range(len(top_disks)):
            for j in range(len(top_disks)):
                if top_disks[i] is not None and (top_disks[j] is None or top_disks[i] < top_disks[j]):
                    moves.append(HanoiMove(i, j))

        return moves

    def get_top_disk(self, tower_index):
        tower = self.get_tower(tower_index)
        if len(tower) > 0:
            return tower[-1]
        return None

    def get_top_disks(self):
        return [self.get_top_disk(i) for i in range(len(self.towers))]

    def get_tower(self, index):
        return self.towers[index]

    def is_complete(self):
        empty_towers = 0
        for tower in self.towers:
            if len(tower) == 0:
                empty_towers = empty_towers + 1
                continue

            break

        return len(self.towers) - empty_towers <= 1


    def __str__(self):
        return "{}".format(self.towers)


class HanoiProblem(Problem):
    def __init__(self):
        super().__init__(default())

    def actions(self, state):
        return state.get_possible_moves()

    def result(self, state, action):
        copy = deepcopy(state)
        copy.apply_move(action)
        return copy

    def goal_test(self, state):
        return state.is_complete()